plugins/weida-raw/fu-weida-raw-device.c

```diff
@@ -0,0 +1,1093 @@
+	FU_WDT8760_CALCULATE_FLASH_CHECKSUM = 0xD5,
+
+} FU_WDT8760Command;
+
+typedef enum {
+	FU_WDT8760_UnprotectLower508k = 0x0044,
+	FU_WDT8760_ProtectAll512k = 0x007C,
+} FU_WDT8760_FlashProtect;
+
+#define FU_WEIDA_RAW_FW_NOT_SUPPORT 0x00
+#define FU_WEIDA_RAW_FW_MAYBE_ISP   0x01
+#define FU_WEIDA_RAW_FW8755	    0x02
+#define FU_WEIDA_RAW_FW8760	    0x4
+#define FU_WEIDA_RAW_FW8762	    0x08
+#define FU_WEIDA_RAW_FW8790	    0X10
+
+struct _FuWeidaRawDevice {
+	FuUdevDevice parent_instance;
+	gint32 dev_type;
+	gint32 firmware_id;
+	gint32 hardware_id;
+	gint32 serial_number;
+	guint8 firmware_rev_ext;
+};
+
+G_DEFINE_TYPE(FuWeidaRawDevice, fu_weida_raw_device, FU_TYPE_UDEV_DEVICE)
+
+static gboolean
+fu_weida_raw_device_set_feature(FuWeidaRawDevice *self,
+				const guint8 *data,
+				guint datasz,
+				GError **error)
+{
+	fu_dump_raw(G_LOG_DOMAIN, "SetFeature", data, datasz);
+	return fu_udev_device_ioctl(FU_UDEV_DEVICE(self),
+				    HIDIOCSFEATURE(datasz),
+				    (guint8 *)data,
+				    NULL,
+				    FU_WEIDA_RAW_DEVICE_IOCTL_TIMEOUT,
+				    error);
+
+	return TRUE;
+}
+
+static gboolean
+fu_weida_raw_device_get_feature(FuWeidaRawDevice *self, guint8 *data, guint datasz, GError **error)
+{
+	if (!fu_udev_device_ioctl(FU_UDEV_DEVICE(self),
+				  HIDIOCGFEATURE(datasz),
+				  data,
+				  NULL,
+				  FU_WEIDA_RAW_DEVICE_IOCTL_TIMEOUT,
+				  error))
+		return FALSE;
+	fu_dump_raw(G_LOG_DOMAIN, "GetFeature", data, datasz);
+
+	return TRUE;
+}
+
+static void
+fu_weida_raw_device_to_string(FuDevice *device, guint idt, GString *str)
+{
+	return;
+}
+
+static gboolean
+fu_weida_raw_device_detach(FuDevice *device, FuProgress *progress, GError **error)
+{
+	return TRUE;
+}
+
+static gboolean
+fu_weida_raw_device_attach(FuDevice *device, FuProgress *progress, GError **error)
+{
+	return TRUE;
+}
+
+static gboolean
+fu_weida_raw_device_reload(FuDevice *device, GError **error)
+{
+	FuWeidaRawDevice *self = FU_WEIDA_RAW_DEVICE(device);
+	/* TODO: reprobe the hardware, or delete this vfunc to use ->setup() as a fallback */
+	g_assert(self != NULL);
+	return TRUE;
+}
+
+static gboolean
+fu_weida_raw_device_probe(FuDevice *device, GError **error)
+{
+	return fu_udev_device_set_physical_id(FU_UDEV_DEVICE(device), "hid", error);
+}
+
+static guint32
+fu_weida_raw_check_firmware_id(FuWeidaRawDevice *self, guint32 fwid)
+{
+	if ((fwid & 0xF0000000) == 0x30000000) {
+		g_message("It is WDT8752 fw !\n");
+		fu_device_set_summary(FU_DEVICE(self), "CI3.0_SR2.0");
+
+		return FU_WEIDA_RAW_FW8755;
+	}
+
+	if ((fwid & 0xFFFF0000) == 0xFFFF0000) {
+		fu_device_set_summary(FU_DEVICE(self), "CI3.0_SR2.0");
+		return FU_WEIDA_RAW_FW8755;
+	}
+
+	if ((fwid & 0xF0000000) == 0x40000000) {
+		fu_device_set_summary(FU_DEVICE(self), "CI4.0_TM4.0");
+		return FU_WEIDA_RAW_FW8760;
+	}
+	if ((fwid & 0xF0000000) == 0x50000000) {
+		fu_device_set_summary(FU_DEVICE(self), "CI5.0");
+		return FU_WEIDA_RAW_FW8790;
+	}
+
+	return FU_WEIDA_RAW_FW_NOT_SUPPORT;
```

Comment:
I think this function should set `self->dev_type = ` and instead return true/false -- if false also setting a GError

Reply:
Done